README.md
<!--
*** Thanks for checking out the Best-README-Template. If you have a suggestion
*** that would make this better, please fork the repo and create a pull request
-->





<!-- PROJECT LOGO -->
<br />
<p align="center">

  <h3 align="center">Flask Lead Creation and Customer Creation API</h3>
</p>



<!-- ABOUT THE PROJECT -->
## About The Project

An application that allows users to create leads, create customers, view leads, edit leads, edit their customers, view customers.

.

Here are some of the features:
* Create Leads
* Create Customers
* Edit Leads
* View Users
* Edit their profiles
* It has implemented DRY principles  :smile:



### Built With

* [Flask](https://flask.palletsprojects.com/en/2.0.x/)
* [Python](https://www.python.org/)
* [PostgresQl](https://www.postgresql.org/)



<!-- GETTING STARTED -->
## Getting Started

This is an example of how you may give instructions on setting up your project locally.
To get a local copy up and running follow these simple example steps.

### Prerequisites

This is an example of how to list things you need to use the software and how to install them.
  ```sh
  
  ```

### Installation

1. Access Github [https://github.com](https://github.com)
2. Clone the repo
   ```sh
   git clone https://github.com/lupamo3/flask-rest-api.git
   ```
3. Change directory into the flask-rest-api directory :
   ```sh
   cd flask-rest-api
   ```
4. Create and activate your virtual environment :

   ```sh
   Virtual venv python=[Python-Version]
   Pip install auto-env
   ```
5. Install Project Requirements
```sh
pip install -r requirements.txt
```
6. Run the application
```sh
flask run
```

### Test the application on Postman
## Test The API end-points
 - Run [CertificateTest](https://farifu-flask.herokuapp.com/) on your postman to test the URLs

or use:

| URL                                 | METHOD                 | MESSAGE                                |
| ------------------------------------|:----------------------:| --------------------------------------:|
|/api/users/register                  | POST                   | Create a leads/customer admin.         |
|/api/users/login                     | POST                   | Login to profile.                      |
|/api/users/all                       | GET                    | Get all users                          |
|/api/users/edit                      | PUT                    | Edit Specific User  records            |
|/api/users/logout                    | POST                   | Logout of the platform                 |
|/api/leads/create                    | POST                   | Create a lead record                   |
|/api/leads/all                       | GET                    | View all leads.                        |
|/api/customer/create                 | POST                   | Create a Customer   .                  |
|/api/customer/all                    | GET                    | Get all customers                      |  




<!-- USAGE EXAMPLES -->
## Additional Information

- **Feel free to reach me via email and to fork this project**
    - Any feedback would be appreciated.
    - The Pull requests have bit by bit application documentation


<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request



<!-- LICENSE -->
## License

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)


<!-- CONTACT -->
## Contact

Your Name - [@nlanjichi](https://twitter.com/nlanjichi)

Project Link: [https://github.com/lupamo3/flask-crud/tree/master](https://github.com/lupamo3/flask-crud/tree/master)




CHANGELOG.md

- Stack: Flask / Flask-JWT-Extended / SQLAlchemy
- API:
   - Sign UP: `/api/users/register`
   - Sign IN: `/api/users/login`
   - Logout: `/api/users/logout`



tests.py

import pytest
import json

from api import app

"""
   Sample test data
"""

DUMMY_USERNAME = "test-basigo"
DUMMY_EMAIL = "test-basigo@basigo.com"
DUMMY_PASS = "basi123go" 

@pytest.fixture
def client():
    with app.test_client() as client:
        yield client

def test_user_signup(client):
    """
       Tests /users/register API
    """
    response = client.post(
        "api/users/register",
        data=json.dumps(
            {
                "username": DUMMY_USERNAME,
                "email": DUMMY_EMAIL,
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 200
    assert "The user was successfully registered" in data["msg"]


def test_user_signup_invalid_data(client):
    """
       Tests /users/register API: invalid data like email field empty
    """
    response = client.post(
        "api/users/register",
        data=json.dumps(
            {
                "username": DUMMY_USERNAME,
                "email": "",
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 400
    assert "'' is too short" in data["msg"]


def test_user_login_correct(client):
    """
       Tests /users/signup API: Correct credentials
    """
    response = client.post(
        "api/users/login",
        data=json.dumps(
            {
                "email": DUMMY_EMAIL,
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 200
    assert data["token"] != ""


def test_user_login_error(client):
    """
       Tests /users/signup API: Wrong credentials
    """
    response = client.post(
        "api/users/login",
        data=json.dumps(
            {
                "email": DUMMY_EMAIL,
                "password": DUMMY_EMAIL
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 400
    assert "Wrong credentials." in data["msg"]


api/models.py

from datetime import datetime

import json
import enum

from werkzeug.security import generate_password_hash, check_password_hash
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy.orm import backref
from flask_login import UserMixin



db = SQLAlchemy()


class Users(UserMixin, db.Model):
    __tablename__ = 'users'

    id = db.Column(db.Integer(), primary_key=True)
    username = db.Column(db.String(32), nullable=False)
    email = db.Column(db.String(64), nullable=False)
    password = db.Column(db.Text())
    jwt_auth_active = db.Column(db.Boolean())
    date_joined = db.Column(db.DateTime(), default=datetime.utcnow)
    roles = db.relationship('Role', secondary='user_roles')
    leads = db.relationship('Leads', backref='users', lazy='dynamic')
    customer = db.relationship('Customer', backref='users', lazy='dynamic')

    def __repr__(self):
        return f"User {self.username}"

    def save(self):
        db.session.add(self)
        db.session.commit()

    def set_password(self, password):
        self.password = generate_password_hash(password)

    def check_password(self, password):
        return check_password_hash(self.password, password)

    def update_email(self, new_email):
        self.email = new_email

    def update_username(self, new_username):
        self.username = new_username

    def check_jwt_auth_active(self):
        return self.jwt_auth_active

    def set_jwt_auth_active(self, set_status):
        self.jwt_auth_active = set_status

    def get_all_users():
        users = Users.query.all()
        return users

    # @login.user_loader
    # def load_user(id):
    #     return User.query.get(int(id))

    @classmethod
    def get_by_id(cls, id):
        return cls.query.get_or_404(id)

    @classmethod
    def get_by_email(cls, email):
        return cls.query.filter_by(email=email).first()

    def toDICT(self):
        cls_dict = {}
        cls_dict['_id'] = self.id
        cls_dict['username'] = self.username
        cls_dict['email'] = self.email
        return cls_dict

    def toJSON(self):
        return self.toDICT()

class RolesDefault(enum.Enum):
    lead_create = '989'
    customer_create = '2' 

# Define the Role data-model
class Role(db.Model):
    __tablename__ = 'roles'
    id = db.Column(db.Integer(), primary_key=True)
    name = db.Column(db.Enum(
                        RolesDefault), 
                        default=RolesDefault.lead_create,
                        nullable=False)

    def __repr__(self):
        return f"Role {self.name}"

    def save(self):
        db.session.add(self)
        db.session.commit()

# Define the UserRoles association table
class UserRoles(db.Model):
    __tablename__ = 'user_roles'
    id = db.Column(db.Integer(), primary_key=True)
    user_id = db.Column(db.Integer(), db.ForeignKey('users.id', ondelete='CASCADE'))
    role_id = db.Column(db.Integer(), db.ForeignKey('roles.id', ondelete='CASCADE'))


class JWTTokenBlocklist(db.Model):
    __tablename__ = 'jwt_token_blocklist'

    id = db.Column(db.Integer(), primary_key=True)
    jwt_token = db.Column(db.String(), nullable=False)
    created_at = db.Column(db.DateTime(), nullable=False)

    def __repr__(self):
        return f"Expired Token: {self.jwt_token}"


    def save(self):
        db.session.add(self)
        db.session.commit()


class Leads(db.Model):

    __tablename__ = 'leads'

    id = db.Column(db.Integer(), primary_key=True)
    first_name = db.Column(db.String(64), nullable=False)
    last_name = db.Column(db.String(64), nullable=False)
    middle_name = db.Column(db.String(64), nullable=False)
    phone_number = db.Column(db.String(64), nullable=False)
    location = db.Column(db.String(64), nullable=False)
    gender = db.Column(db.String(64), nullable=False)
    date_created = db.Column(db.DateTime(), default=datetime.utcnow)
    created_by = db.Column(db.Integer, db.ForeignKey('users.id'))
    # created_by_user = db.relationship("Users", backref=backref("leads", uselist=False))


    def __repr__(self):
        return f"Leads {self.first_name} {self.last_name}"

    def save(self):
        db.session.add(self)
        db.session.commit()

    def toDICT(self):
        cls_dict = {}
        cls_dict['_id'] = self.id
        cls_dict['first_name'] = self.first_name
        cls_dict['last_name'] = self.last_name
        cls_dict['middle_name'] = self.middle_name
        cls_dict['phone_number'] = self.phone_number
        cls_dict['location'] = self.location
        cls_dict['gender'] = self.gender
        cls_dict['date_created'] = self.date_created
        cls_dict['created_by'] = self.created_by
        # cls_dict['created_by_user'] = self.created_by_user

    def toJSON(self):
        return self.toDICT()

    def get_by_id(cls, id):
        return cls.query.get_or_404(id)

    def get_all_leads():
        leads = Leads.query.all()
        return leads

class ProductsofInterest(enum.Enum):
    product_a = 10,000
    product_b = 20,000  
    product_c = 30,000

class Customer(db.Model):
    __tablename__ = 'customer'
    

    id = db.Column(db.Integer(), primary_key=True)
    photo = db.Column(db.String(64), nullable=False)
    annual_earnings = db.Column(db.String(64), nullable=False)
    date_created = db.Column(db.DateTime(), default=datetime.utcnow)
    products_interested = db.Column(
                                db.Enum(ProductsofInterest), 
                                default=ProductsofInterest.product_a, 
                                nullable=False
    )
    created_by = db.Column(db.Integer, db.ForeignKey('users.id'))

    def __repr__(self):
        return f"Customer {self.id}"

    def save(self):
        db.session.add(self)
        db.session.commit()

    def toDICT(self):
        cls_dict = {}
        cls_dict['_id'] = self.id
        cls_dict['photo'] = self.photo
        cls_dict['annual_earnings'] = self.annual_earnings
        cls_dict['date_created'] = self.date_created
        cls_dict['products_interested'] = self.products_interested
        cls_dict['created_by'] = self.created_by
        return cls_dict

    def get_by_id(cls, id):
        return cls.query.get_or_404(id)

    def toJSON(self):
        return self.toDICT()

    def update_photo(self, new_photo):
        self.photo = new_photo

    def get_all_customers():
        customers = Customer.query.all()
        print("customers", customers)
        return customers

    def get_specific_customer(products_interested):
        customer = Customer.query.filter_by(products_interested='product_a')
        return customer

    




api/__init__.py

import json

from flask import Flask
from flask_cors import CORS
from flask_login import LoginManager

from .routes import rest_api
from .models import db

app = Flask(__name__)
login = LoginManager(app)


app.config.from_object('api.config.BaseConfig')

db.init_app(app)
rest_api.init_app(app)
CORS(app)

# Setup database
@app.before_first_request
def initialize_database():
    db.create_all()

"""
   Custom responses
"""

@app.after_request
def after_request(response):
    """
       Sends back a custom error with {"success", "msg"} format
    """

    if int(response.status_code) >= 400:
        response_data = json.loads(response.get_data())
        if "errors" in response_data:
            response_data = {"success": False,
                             "msg": list(response_data["errors"].items())[0][1]}
            response.set_data(json.dumps(response_data))
        response.headers.add('Content-Type', 'application/json')
    return response


api/routes.py
from datetime import datetime, timezone, timedelta

from functools import wraps

from flask import request
from flask_restx import Api, Resource, fields

import jwt

from .models import db, Users, JWTTokenBlocklist, Leads, Customer
from .config import BaseConfig

rest_api = Api(version="1.0", title="Users API")


"""
    Flask-Restx models for api request and response data
"""

signup_model = rest_api.model(
    "SignUpModel",
    {
        "username": fields.String(required=True, min_length=2, max_length=32),
        "email": fields.String(required=True, min_length=4, max_length=64),
        "password": fields.String(required=True, min_length=4, max_length=16),
    },
)

login_model = rest_api.model(
    "LoginModel",
    {
        "email": fields.String(required=True, min_length=4, max_length=64),
        "password": fields.String(required=True, min_length=4, max_length=16),
    },
)

user_edit_model = rest_api.model(
    "UserEditModel",
    {
        "userID": fields.String(required=True, min_length=1, max_length=32),
        "username": fields.String(required=True, min_length=2, max_length=32),
        "email": fields.String(required=True, min_length=4, max_length=64),
    },
)
lead_model = rest_api.model(
    "LeadModel",
    {
        "first_name": fields.String(required=True, min_length=2, max_length=64),
        "last_name": fields.String(required=True, min_length=2, max_length=64),
        "middle_name": fields.String(required=True, min_length=2, max_length=64),
        "phone_number": fields.String(required=True, min_length=2, max_length=64),
        "location": fields.String(required=True, min_length=4, max_length=64),
        "gender": fields.String(required=True, min_length=2, max_length=64),
    },
)

customer_model = rest_api.model(
    "CustomerModel",
    {
        "photo": fields.String(required=True, min_length=2, max_length=64),
        "annual_earnings": fields.String(required=True, min_length=2, max_length=64),
        "products_interested": fields.String(
            required=True, min_length=2, max_length=64
        ),
    },
)


"""
   Helper function for JWT token required
"""


def token_required(f):
    @wraps(f)
    def decorator(*args, **kwargs):

        token = None

        if "authorization" in request.headers:
            token = request.headers["authorization"]

        if not token:
            return {"success": False, "msg": "Valid JWT token is missing"}, 400

        try:
            data = jwt.decode(token, BaseConfig.SECRET_KEY, algorithms=["HS256"])
            current_user = Users.get_by_email(data["email"])

            if not current_user:
                return {
                    "success": False,
                    "msg": "Sorry. Wrong auth token. This user does not exist.",
                }, 400

            token_expired = (
                db.session.query(JWTTokenBlocklist.id)
                .filter_by(jwt_token=token)
                .scalar()
            )

            if token_expired is not None:
                return {"success": False, "msg": "Token revoked."}, 400

            if not current_user.check_jwt_auth_active():
                return {"success": False, "msg": "Token expired."}, 400

        except:
            return {"success": False, "msg": "Token is invalid"}, 400

        return f(current_user, *args, **kwargs)

    return decorator


"""
    Flask-Restx routes
"""


@rest_api.route("/api/users/register")
class Register(Resource):
    """
    Creates a new user by taking 'signup_model' input
    """

    @rest_api.expect(signup_model, validate=True)
    def post(self):

        req_data = request.get_json()

        _username = req_data.get("username")
        _email = req_data.get("email")
        _password = req_data.get("password")

        user_exists = Users.get_by_email(_email)
        if user_exists:
            return {"success": False, "msg": "Email already taken"}, 400

        new_user = Users(username=_username, email=_email)

        new_user.set_password(_password)
        new_user.save()

        return {
            "success": True,
            "userID": new_user.id,
            "msg": "The user was successfully registered",
        }, 200


@rest_api.route("/api/users/login")
class Login(Resource):
    """
    Login user by taking 'login_model' input and return JWT token
    """

    @rest_api.expect(login_model, validate=True)
    def post(self):

        req_data = request.get_json()

        _email = req_data.get("email")
        _password = req_data.get("password")

        user_exists = Users.get_by_email(_email)

        if not user_exists:
            return {"success": False, "msg": "This email does not exist."}, 400

        if not user_exists.check_password(_password):
            return {"success": False, "msg": "Wrong credentials."}, 400

        # create access token uwing JWT
        token = jwt.encode(
            {"email": _email, "exp": datetime.utcnow() + timedelta(minutes=30)},
            BaseConfig.SECRET_KEY,
        )

        user_exists.set_jwt_auth_active(True)
        user_exists.save()

        return {"success": True, "token": token, "user": user_exists.toJSON()}, 200


@rest_api.route("/api/users/all")
class GetAllUsers(Resource):
    """
    Returns all users
    """

    @token_required
    def get(current_user):

        users = Users.get_all_users()
        return {"success": True, "users": [user.toJSON() for user in users]}, 200


@rest_api.route("/api/users/edit")
class EditUser(Resource):
    """
    Edits User's username or password or both using 'user_edit_model' input
    """

    @rest_api.expect(user_edit_model)
    @token_required
    def post(self, current_user):

        req_data = request.get_json()

        _new_username = req_data.get("username")
        _new_email = req_data.get("email")

        if _new_username:
            self.update_username(_new_username)

        if _new_email:
            self.update_email(_new_email)

        self.save()

        return {"success": True}, 200


@rest_api.route("/api/users/logout")
class LogoutUser(Resource):
    """
    Logs out User using 'logout_model' input
    """

    @token_required
    def post(self, current_user):

        _jwt_token = request.headers["authorization"]

        jwt_block = JWTTokenBlocklist(
            jwt_token=_jwt_token, created_at=datetime.now(timezone.utc)
        )
        jwt_block.save()

        self.set_jwt_auth_active(False)
        self.save()

        return {"success": True}, 200


@rest_api.route("/api/leads/create")
class CreateLead(Resource):
    """
    Create a Lead
    """

    @rest_api.expect(lead_model, validate=True)
    @token_required
    def post(self, current_user):

        req_data = request.get_json()

        _first_name = req_data.get("first_name")
        _last_name = req_data.get("last_name")
        _middle_name = req_data.get("middle_name")
        _phone_number = req_data.get("phone_number")
        _gender = req_data.get("gender")
        _location = req_data.get("location")

        # lead_exists = Leads.get_by_id()

        # if not lead_exists:
        #     return {"success": False,
        #             "msg": "This lead does not exist."}, 400

        new_lead = Leads(
            first_name=_first_name,
            last_name=_last_name,
            middle_name=_middle_name,
            phone_number=_phone_number,
            gender=_gender,
            location=_location,
        )
        new_lead.save()

        return {
            "success": True,
            "msg": "The lead was successfully created",
        }, 200


@rest_api.route("/api/leads/all")
class GetAllLeads(Resource):
    """
    Returns all leads
    """

    @token_required
    def get(self, current_user):

        leads = Leads.get_all_leads()
        return {"success": True, "leads": [lead.toJSON() for lead in leads]}, 200


@rest_api.route("/api/customer/create")
class CreateCustomer(Resource):
    """
    Create a Customer
    """

    @rest_api.expect(customer_model, validate=True)
    @token_required
    def post(self, current_user):
        req_data = request.get_json()

        _photo = req_data.get("photo")
        _annual_earnings = req_data.get("annual_earnings")
        _products_interested = req_data.get("products_interested")

        new_customer = Customer(
            photo=_photo,
            annual_earnings=_annual_earnings,
            products_interested=_products_interested,
        )
        new_customer.save()

        return {"success": True}, 200

@rest_api.route("/api/customer/all")
class GetAllCustomers(Resource):
    """
    Returns all customers
    """

    @token_required
    def get(self, current_user):
        customer = Customer.get_specific_customer('product_a')
        print('SHOW CUST', customer)
        return {"success": True, "customer": customer}, 200
